### ISSUE TITLE
proposal: crane: add the ability to remove labels and annotations through mutate command

### ISSUE THREAD

**Issue by Dentrax** [google/go-containerregistry#1062]
**TL;DR:** I propose adding a new flag called `--remove` _(`-r` for shorthand)_ for `mutate` command to remove _existing_ label and annotations.

**Proposal**
If we pass `--remove` flag to `mutate` command, the actual flow will be changed. Currently, we [check](https://github.com/google/go-containerregistry/blob/ce35c99b3fafc7b6ea9eb192933ae2730fe360d0/cmd/crane/cmd/mutate.go#L133) if a given _value_ is in `key=value` format. If we want to remove a _label_ or an _annotation_, we really do not want to pass its _value_, since each key is unique.

**Implementation**
At a minimum, `--remove` flag will impact:
* We can not pass a value to this flag
* It's just a _bool_ value, if passed, we will in _removeMode_
* This flag traverses all the given _label_s and _annotation_s

**Example**

Some happy paths are the following commands:

* Initially, mutate with some key and values:
```
$ crane mutate --label foo=bar --label bar=baz --annotation qux=quux $IMAGE
```

* Removing the _existing_ label with a given _key_
```
$ crane mutate --remove --label foo $IMAGE # OK
```

* Removing the _existing_ label and annotation with given _keys_
```
$ crane mutate --remove --label foo --annotation qux $IMAGE # OK
```

* Removing the _multiple-existing_ labels with given _keys_
```
$ crane mutate --remove --label foo --label bar $IMAGE # OK
```

* Removing the _non-existing_ label with a given _key_
```
$ crane mutate --remove --label qux $IMAGE # OH!
```

* Removing the _non-existing_ label with a given _key=value_
```
$ crane mutate --remove --label qux=quux $IMAGE # Not allowed!
```

* Removing the _existing_ label with a given _key=value_
```
$ crane mutate --remove --label bar=baz $IMAGE # Not allowed!
```

**P.S:** All the given instruction steps for the `--label` _(stated in the examples)_, will accurately the same for the `--annotation` flag.

**Additional Notes**

If you ask that why not this should be in the `delete` command, I am not so sure about that idea since we do not "apply" any _mutation_ operation there. It is just for _deletion_ for the given image.

**Comment by Dentrax** [comment]
cc: @developer-guy

**Comment by jonjohnsonjr** [comment]
I think I'd rather see these as `--unlabel` and `--unannotate` so that you can mix and match things.

I want to eventually use the `-r` and `--rm` flags for other reasons, so having this be a global flag that inverts the behavior feels confusing.

**Comment by Dentrax** [comment]
Hmm, actually, I thought that passing two different flags through `mutate` command may complicate DX a bit. Like the following cases:

* This command will remove both `foo` and `bar` labels:
```
$ crane mutate --unlabel --label foo --label bar
```

* This command will remove `foo` label and will add `qux` annotation:
_(with single command, we added and removed "things" at the same time)_
I am not sure if usage provides a decent developer experience.
```
$ crane mutate --unlabel --label foo --annotation qux=quux
```

Additionally, we _may_ have to split the `splitKeyVals` function as we bring two new boolean flags.

```
$ crane mutate --unlabel --label foo --annotation qux=quux
```
^^^ For this scenario, _labels_ should NOT be passed in `key=value` format, since it is OK to pass for _annotation_ flags in that format, and vice versa.

**P.S:** If you have future plans to use `-r` and `--rm` flags, we can simply pass `-d, --delete` flag either. But this may bring some confusing with `--debug` flag, it that planned too.

**Comment by imjasonh** [comment]
I was imagining something along the lines of:

```
$ crane mutate --label foo=bar --annotate baz=quux  # adds labels and annotations
$ crane mutate --unlabel foo --unannotate baz       # removes labels and annotations
```

If it's passed `--label foo=bar --unlabel foo` it's an error. If passed `--unlabel foo` but the image doesn't have a label for `foo` it's a no-op.

Both `--unlabel` and `--unannotate` can be passed multiple times to remove multiple values, like how `--label` and `--annotate` can be passed multiple times to add multiple values.

This seems a bit easier to grok than having an "unlabel mode" that semantically inverts `--label` flag values.

**Comment by Dentrax** [comment]
Oh, I see! I got it wrong that logic, I thought that it was act a `boolean` "mode". Thanks for the clarification and feedbacks. We (@developer-guy) will try to get this into soon.

**Comment by imjasonh** [comment]
I'm curious, do you have a practical use case in mind for this? In my experience things rarely set/update labels/annotations dynamically (vs in some config like a Dockerfile), and even more rarely need to remove them on an image that has them set. I can't think of a single time I've wanted to remove a label or annotation from an image, vs just rebuilding it without the label/annotation next time.

Since there are some practical implementation and UX questions floating around this issue, I want to make sure we're solving a real world problem, so we can consult the real prospective users about what their preferred usage would be.
